warehouse/src/main/java/org/dromara/hertzbeat/warehouse/store/HistoryGrepTimeDbDataStorage.java

```diff
@@ -0,0 +1,407 @@
+			} catch (InterruptedException | ExecutionException e) {
+				log.error("Error creating database");
+			}
+		}
+		return isDatabaseExist;
+	}
+
+	@Override
+	void saveData(CollectRep.MetricsData metricsData) {
+		if (!isServerAvailable() || metricsData.getCode() != CollectRep.Code.SUCCESS) {
+			return;
+		}
+		if (metricsData.getValuesList().isEmpty()) {
+			log.info("[warehouse greptime] flush metrics data {} is null, ignore.", metricsData.getId());
+			return;
+		}
+		String monitorId = String.valueOf(metricsData.getId());
+		String table = metricsData.getApp() + "_" + metricsData.getMetrics();
+		//TODO bug：选择STORAGE_DATABASE不起作用，还是默认存在public里
+		TableSchema.Builder tableSchemaBuilder = TableSchema.newBuilder(TableName.with(STORAGE_DATABASE, table));
+
+		List<SemanticType> semanticTypes = new LinkedList<>(Arrays.asList(SemanticType.Tag, SemanticType.Tag, SemanticType.Timestamp));
+		List<ColumnDataType> dataTypes = new LinkedList<>(Arrays.asList(ColumnDataType.String, ColumnDataType.String, ColumnDataType.Int64));
+		List<String> columnNames = new LinkedList<>(Arrays.asList("monitor_id", "instance", "ts"));
+
+		List<CollectRep.Field> fieldsList = metricsData.getFieldsList();
+		for (CollectRep.Field field : fieldsList) {
+			semanticTypes.add(SemanticType.Field);
+			columnNames.add(field.getName());
+			// handle field type
+			if (field.getType() == CommonConstants.TYPE_NUMBER) {
+				dataTypes.add(ColumnDataType.Float64);
+			} else if (field.getType() == CommonConstants.TYPE_STRING) {
+				dataTypes.add(ColumnDataType.String);
+			}
+		}
+		tableSchemaBuilder.semanticTypes(semanticTypes.toArray(new SemanticType[0]));
+		tableSchemaBuilder.dataTypes(dataTypes.toArray(new ColumnDataType[0]));
+		tableSchemaBuilder.columnNames(columnNames.toArray(new String[0]));
+		WriteRows rows = WriteRows.newBuilder(tableSchemaBuilder.build()).build();
+		try {
+			long now = System.currentTimeMillis();
+			Object[] values = new Object[3 + fieldsList.size()];
+			values[0] = monitorId;
+			values[2] = now;
+			for (CollectRep.ValueRow valueRow : metricsData.getValuesList()) {
+				String instance = valueRow.getInstance();
+				if (!instance.isEmpty()) {
+					instance = String.format("\"%s\"", instance);
+					values[1] = instance;
+				} else {
+					values[1] = null;
+				}
+				for (int i = 0; i < fieldsList.size(); i++) {
+					if (!CommonConstants.NULL_VALUE.equals(valueRow.getColumns(i))) {
+						if (fieldsList.get(i).getType() == CommonConstants.TYPE_NUMBER) {
+							values[3 + i] = Double.parseDouble(valueRow.getColumns(i));
+						} else if (fieldsList.get(i).getType() == CommonConstants.TYPE_STRING) {
+							values[3 + i] = valueRow.getColumns(i);
+						}
+					} else {
+						values[3 + i] = null;
+					}
+				}
+				rows.insert(values);
+			}
+			rows.finish();
+			CompletableFuture<Result<WriteOk, Err>> writeFuture = greptimeDb.write(rows);
+			writeFuture.whenComplete((result, throwable) -> {
```

Comment:
I think we should wait here for the `writeFuture` to be completed, otherwise the results may be lost.